File: packages/api-models-ts/src/Filters.ts
/**
 * Request and response models for `/api/filters`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Schemas from "./Schemas.ts";

/**
 * An alert-stream filter belonging to a group (upstream `Filter`).
 *
 * `stream`, `group`, `broker` and `candidates` stay untyped: each of those
 * upstream models owns a `filters` (or `filter`) relationship, so typing them
 * here would risk an import cycle.
 *
 * @since 1.0.0
 * @category Models
 */
export const Filter = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        name: Schemas.NullishString,
        stream_id: Schemas.NullishInteger,
        group_id: Schemas.NullishInteger,
        broker_id: Schemas.NullishInteger,
        altdata: Schemas.nullish(Schemas.JsonObject),
        autosave: Schemas.NullishBoolean,
        stream: Schemas.nullish(Schemas.JsonObject),
        group: Schemas.nullish(Schemas.JsonObject),
        broker: Schemas.nullish(Schemas.JsonObject),
        candidates: Schemas.nullish(v.array(Schemas.JsonObject)),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Filter = v.InferOutput<typeof Filter>;

/**
 * Payload for creating a filter.
 *
 * @since 1.0.0
 * @category Models
 */
export interface FilterPost {
    readonly name: string;
    readonly stream_id: number;
    readonly group_id: number;
    /** The broker the filter runs on, if any. */
    readonly broker_id?: number | undefined;
    /** Arbitrary extra JSON. */
    readonly altdata?: Record<string, unknown> | undefined;
}

/**
 * Payload for updating a filter.
 *
 * @since 1.0.0
 * @category Models
 */
export interface FilterPatch {
    readonly name?: string | undefined;
    readonly altdata?: Record<string, unknown> | undefined;
    readonly group_id?: number | undefined;
    readonly stream_id?: number | undefined;
    readonly autosave?: boolean | undefined;
}

/**
 * Result of creating a filter.
 *
 * @since 1.0.0
 * @category Models
 */
export const FilterPostResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type FilterPostResponse = v.InferOutput<typeof FilterPostResponse>;

File: packages/api-models-py/skyportal_py_models/filters.py
"""Request and response models for ``/api/filters``."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class Filter(BaseModel):
    """An alert-stream filter belonging to a group (upstream ``Filter``).

    ``stream``, ``group``, ``broker`` and ``candidates`` stay untyped: each of
    those upstream models owns a ``filters`` (or ``filter``) relationship, so
    typing them here would risk an import cycle.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime | None = None
    modified: datetime | None = None
    name: str | None = None
    stream_id: int | None = None
    group_id: int | None = None
    broker_id: int | None = None
    altdata: dict[str, Any] | None = None
    autosave: bool | None = None
    stream: dict[str, Any] | None = None
    group: dict[str, Any] | None = None
    broker: dict[str, Any] | None = None
    candidates: list[dict[str, Any]] | None = None


class FilterPost(BaseModel):
    """Payload for creating a filter."""

    model_config = ConfigDict(extra="forbid")

    name: str
    stream_id: int
    group_id: int
    broker_id: int | None = None
    altdata: dict[str, Any] | None = None


class FilterPatch(BaseModel):
    """Payload for updating a filter."""

    model_config = ConfigDict(extra="forbid")

    name: str | None = None
    altdata: dict[str, Any] | None = None
    group_id: int | None = None
    stream_id: int | None = None
    autosave: bool | None = None


class FilterPostResponse(BaseModel):
    """Result of creating a filter."""

    model_config = ConfigDict(extra="forbid")

    id: int
